### PR TITLE
Add CloseTextSampler and FixedIndicesSampler

### PR DESCRIPTION
**Pull request by yoavkatz**
CloseTextSampler -  returns demos that are textually close to the current instance.
FixedIndicesSampler - return fixed demos based on indices in the train pool

Also fixed a bug in reproducibility of demos by making sure randomization is based on instance, and being global (so it is shared with ther instances)

### CODE REVIEW

File: src/unitxt/splitters.py [hunk 1]
```diff
@@ -151,13 +153,52 @@ def filter_source_by_instance(
...
+        )
+        # Randmly select 'sample_size' instances that are from the closest matches text
+        # (There may be multiple instance with same text in the given field, and the order returned is
+        # is also randomized )
+        instances_pool = [
+            instance_in_pool
+            for instance_in_pool in instances_pool
+            if dict_get(instance_in_pool, field) in closest_matches
+        ]
+        return self.random_generator.sample(instances_pool, self.sample_size)
```
Comment (elronbandel) [inline src/unitxt/splitters.py:199]: Why sampling is needed? isn't a list of size sample_size by now?

File: src/unitxt/splitters.py [hunk 2]
```diff
@@ -151,13 +153,52 @@ def filter_source_by_instance(
...
 
 
+class CloseTextSampler(Sampler):
+    """Selects the samples of instances which are the closest textual match to the given instance.
+
+    Comparison is done based on a given field in the instance.
+
+    """
+
+    field: str
```
Comment (elronbandel) [inline src/unitxt/splitters.py:174]: Maybe have a None default that uses just the serialization of the inputs?

**Review by elronbandel — CHANGES_REQUESTED** [review]
See comments

File: src/unitxt/splitters.py (hunk 1 above)
Comment (yoavkatz) [reply src/unitxt/splitters.py:199]: The reason is noted above:
# (There may be multiple instance with same text in the given field, and the order returned is
# is also randomized )

File: src/unitxt/splitters.py (hunk 2 above)
Comment (yoavkatz) [reply src/unitxt/splitters.py:174]: This seems bit of an odd default.  Consider inputs such as "text_type" or "classes" , which you don't want to take part in the the calculation.

Comment (elronbandel) [reply src/unitxt/splitters.py:174]: The do not harm either. Anyways just a suggestion

**Review by elronbandel — APPROVED** [review]

**Comment by elronbandel** [comment]
Relevant to https://github.com/IBM/unitxt/issues/948

File: src/unitxt/splitters.py [hunk 3]
```diff
@@ -151,11 +146,74 @@ def filter_source_by_instance(
...
+
     def sample(
-        self, instances_pool: List[Dict[str, object]]
+        self,
+        instances_pool: List[Dict[str, object]],
+        instance: Optional[Dict[str, object]] = None,
     ) -> List[Dict[str, object]]:
         instances_pool = list(instances_pool)
-        return self.random_generator.sample(instances_pool, self.sample_size)
+        random_generator = new_random_generator(sub_seed=str(instance))
```
Comment (elronbandel) [inline src/unitxt/splitters.py:157]:
```suggestion
        random_generator = new_random_generator(sub_seed={**instance["inputs"], **instance["outputs"]})
```
Make it more robust to future changes in the instance structure (such as classification policy metadata etc that does not change the essence of the instance)

**Review by elronbandel — DISMISSED** [review]
Looks like really important changes. Should we add to the metadata the indices of chosen demos in all the cases? It can help people reproduce or just locate good demos in the future.

File: src/unitxt/splitters.py (hunk 3 above)
Comment (yoavkatz) [reply src/unitxt/splitters.py]: Thanks.  I changed to code to use "input_fields" (references are not always available, and should not effect randomization if the input fields are the same).

**Comment by yoavkatz** [comment]
> Looks like really important changes. Should we add to the metadata the indices of chosen demos in all the cases? It can help people reproduce or just locate good demos in the future.

I'm not sure this is a required use case.    For most uses, there are different demos per example, so when you average the score, you average over different indices.  You don't really need to know which examples instance indices were chosen for each text.

**Comment by dafnapension** [comment]
Hi @yoavkatz  and @elronbandel ,  I think this is very nice, and understand developed per needs raised, and certainly solves the inconsistency problem.

In my branch, I used the occasion to remove a rather strange class, I think, `InstanceOperatorWithMultiStreamAccess`  which was only used for `SpreadSplit`,  and implemented `SpreadSplit` as a `MultiStreamOperator`.  Would you like to add this change also here?

**Comment by yoavkatz** [comment]
> Hi @yoavkatz and @elronbandel , I think this is very nice, and understand developed per needs raised, and certainly solves the inconsistency problem.
> 
> In my branch, I used the occasion to remove a rather strange class, I think, `InstanceOperatorWithMultiStreamAccess` which was only used for `SpreadSplit`, and implemented `SpreadSplit` as a `MultiStreamOperator`. Would you like to add this change also here?

I think while right now this is the SpreadSplit is the only class that uses InstanceOperatorWithMultiStreamAccess , it may be useful for other classes in the future.

**Pull request merged**